### PR TITLE
JBIDE-19056: A mojo to check version

### PR DESCRIPTION
**Pull request by mickaelistria**

### CODE REVIEW

**Comment by nickboldt** [comment]
+1 to apply, but see comments in https://github.com/jbosstools/jbosstools-build/pull/172/ and https://issues.jboss.org/browse/JBIDE-19056 ... might be a problem or two and one thing that can be improved (move baseline URL to a property so easier to maintain).

**Comment by maxandersen** [comment]
I tried but I couldn't reproduce the issues @nickboldt  had about many warnings etc. mentioned in jbosstools/jbosstools-build#172

File: tycho-plugins/repository-utils/src/main/java/org/jboss/tools/tycho/sitegenerator/CompareWithBaselineMojo.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+	public void execute() throws MojoExecutionException, MojoFailureException {
+		if (skip) {
+			getLog().info("Skipped");
+			return;
+		}
+		ReactorProject reactorProject = DefaultReactorProject.adapt(project);
+		Set<?> dependencyMetadata = reactorProject.getDependencyMetadata(true);
+		if (dependencyMetadata == null || dependencyMetadata.isEmpty()) {
+			getLog().info("No p2 artifacts created");
```
Comment (maxandersen) [inline tycho-plugins/repository-utils/src/main/java/org/jboss/tools/tycho/sitegenerator/CompareWithBaselineMojo.java]: This gets printed exactly on all the projects that does *not* need to know to have any checks made.

I suggest we either add a verbose="true" to enable this or only have it shown when running with -X.

I also suggest to make the message more informative/explicit, i.e. "No p2 artifacts created in build. Skipping p2 baseline version comparison"

File: tycho-plugins/repository-utils/src/main/java/org/jboss/tools/tycho/sitegenerator/CompareWithBaselineMojo.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+				Method getVersionMethod = item.getClass().getMethod("getVersion");
+				String id = (String)getIdMethod.invoke(item);
+				Version version = new Version(getVersionMethod.invoke(item).toString());
+				P2ResolutionResult res = resolver.resolveInstallableUnit(baselineTP, id, "0.0.0");
+
+				for (Entry foundInBaseline : res.getArtifacts()) {
+					Version baselineVersion = new Version(foundInBaseline.getVersion());
+					if (version.compareTo(baselineVersion) <= 0) {
+						throw new MojoFailureException("Version of '" + id + '/' + version
+								+ ") must be bigger than baseline one (" + baselineVersion.toString() + ")");
```
Comment (maxandersen) [inline tycho-plugins/repository-utils/src/main/java/org/jboss/tools/tycho/sitegenerator/CompareWithBaselineMojo.java]: This reads a bit weird in the console.

Suggestion for more explicit messages as at https://github.com/maxandersen/jbosstools-maven-plugins/commit/70bb67bd3f297215e1e834557eb269c13d5841c1

File: tycho-plugins/repository-utils/src/main/java/org/jboss/tools/tycho/sitegenerator/CompareWithBaselineMojo.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+		P2ResolverFactory resolverFactory = this.equinox.getService(P2ResolverFactory.class);
+		P2Resolver resolver = resolverFactory.createResolver(new MavenLoggerAdapter(this.plexusLogger, true));
+
+		TargetPlatformConfigurationStub baselineTPStub = new TargetPlatformConfigurationStub();
+		baselineTPStub.setEnvironments(Collections.singletonList(TargetEnvironment.getRunningEnvironment()));
+		for (String baselineRepo : this.baselines) {
+			baselineTPStub.addP2Repository(toRepoURI(baselineRepo));
+		}
+		TargetPlatform baselineTP = resolverFactory.getTargetPlatformFactory().createTargetPlatform(baselineTPStub, new ExecutionEnvironmentConfigurationStub("JavaSE-1.8"), null, null);
```
Comment (maxandersen) [inline tycho-plugins/repository-utils/src/main/java/org/jboss/tools/tycho/sitegenerator/CompareWithBaselineMojo.java]: why hardcoded java 1.8?  should it not use the environment specified in the build ?

File: tycho-plugins/repository-utils/src/main/java/org/jboss/tools/tycho/sitegenerator/CompareWithBaselineMojo.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+						// TODO: currently, there are only 2 kinds of (known) artifacts, but we could have more
+						// and unknown ones. Need to find something smarter to map artifact with actual file.
+						if (id.endsWith("source")) {
+							currentFile = reactorProject.getArtifact("sources");
+						} else {
+							currentFile = reactorProject.getArtifact();
+						}
+						if (!FileUtils.contentEquals(currentFile, baselineFile)) {
+							throw new MojoFailureException("An artifact with same id and version (" + id + "/" + version + ") exists in baseline. This is not legal, unless they are exactly same artifacts."
+								+ " TODO: compare jars.");
```
Comment (maxandersen) [inline tycho-plugins/repository-utils/src/main/java/org/jboss/tools/tycho/sitegenerator/CompareWithBaselineMojo.java]: TODO still relevant ?

File: tycho-plugins/repository-utils/src/main/java/org/jboss/tools/tycho/sitegenerator/CompareWithBaselineMojo.java (hunk 3 above)
Comment (mickaelistria) [inline tycho-plugins/repository-utils/src/main/java/org/jboss/tools/tycho/sitegenerator/CompareWithBaselineMojo.java]: Right, that'd be better.

File: tycho-plugins/repository-utils/src/main/java/org/jboss/tools/tycho/sitegenerator/CompareWithBaselineMojo.java (hunk 4 above)
Comment (mickaelistria) [inline tycho-plugins/repository-utils/src/main/java/org/jboss/tools/tycho/sitegenerator/CompareWithBaselineMojo.java]: Nope. I missed this one ;)

File: tycho-plugins/repository-utils/src/main/java/org/jboss/tools/tycho/sitegenerator/CompareWithBaselineMojo.java (hunk 3 above)
Comment (mickaelistria) [inline tycho-plugins/repository-utils/src/main/java/org/jboss/tools/tycho/sitegenerator/CompareWithBaselineMojo.java]: We can use TychoProjectsUtils.getExecutionEnvironment(this.project) here.

**Comment by mickaelistria** [comment]
Rebased and merged as 9d7ec0e6b120c2afff02cfadf362912163bf3b50 , with improvements in ac308a9a436041c332c48eae786116aa6318cb01

**Pull request closed without merging**